crates/runtime/src/component/resources.rs

```diff
@@ -294,36 +298,42 @@ impl ResourceTable {
             Slot::Free { next } => next,
             _ => unreachable!(),
         };
-        u32::try_from(ret).unwrap()
+
+        // The component model reserves index 0 as never allocatable so add one
+        // to the table index to start the numbering at 1 instead. Also note
+        // that the component model places an upper-limit per-table on the
+        // maximum allowed index.
+        let ret = ret + 1;
+        if ret >= 1 << 30 {
```

Comment:
Nit: can the upper limit be moved to a constant and perhaps link to the spec where it is specified?